lib/matplotlib/testing/compare.py

```diff
@@ -141,38 +137,81 @@ def _shlex_quote_bytes(b):
-        # other modules (e.g. `subprocess`) may already have their globals
-        # set to `None`, which make `proc.communicate` or `proc.terminate`
-        # fail.  By relying on `atexit` we ensure the destructor runs before
-        # `None`-setting occurs.
+        # Explicitly register deletion from an atexit handler because if we
+        # wait until the object is GC'd (which occurs later), then some module
+        # globals (e.g. signal.SIGKILL) has already been set to None, and
+        # kill() doesn't work anymore...
         atexit.register(self.__del__)
 
-    def _read_to_prompt(self):
-        """Did Inkscape reach the prompt without crashing?
-        """
-        stream = iter(functools.partial(self._proc.stdout.read, 1), b"")
-        prompt = (b"\n", b">")
-        n = len(prompt)
-        its = itertools.tee(stream, n)
-        for i, it in enumerate(its):
-            next(itertools.islice(it, i, i), None)  # Advance `it` by `i`.
+    def __del__(self):
+        if self._proc:
+            self._proc.kill()
+            self._proc.wait()
+            for stream in filter(None, [self._proc.stdin,
+                                        self._proc.stdout,
+                                        self._proc.stderr]):
+                stream.close()
+            self._proc = None
+
+    def _read_until(self, terminator):
+        """Read until the prompt is reached."""
+        buf = bytearray()
         while True:
-            window = tuple(map(next, its))
-            if len(window) != n:
-                # Ran out of data -- one of the `next(it)` raised
-                # StopIteration, so the tuple is shorter.
-                return False
-            if self._proc.poll() is not None:
-                # Inkscape exited.
-                return False
-            if window == prompt:
-                # Successfully read until prompt.
-                return True
+            c = self._proc.stdout.read(1)
+            if not c:
+                raise _ConverterError
+            buf.extend(c)
+            if buf.endswith(terminator):
+                return bytes(buf[:-len(terminator)])
 
+
+class _GSConverter(_Converter):
+    def __call__(self, orig, dest):
+        if not self._proc:
+            self._stdout = TemporaryFile()
+            self._proc = subprocess.Popen(
+                [matplotlib.checkdep_ghostscript.executable,
+                 "-dNOPAUSE", "-sDEVICE=png16m"],
+                # As far as I can see, ghostscript never outputs to stderr.
+                stdin=subprocess.PIPE, stdout=subprocess.PIPE)
+            try:
+                self._read_until(b"\nGS")
+            except _ConverterError:
+                raise OSError("Failed to start Ghostscript")
+
+        def encode_and_escape(name):
+            return (os.fsencode(name)
+                    .replace(b"(", br"\(")
+                    .replace(b")", br"\)")
+                    .replace(b"\\", b"\\\\"))
```

Comment:
Shouldn't the backslash escaping come first to avoid breaking the escaping of parentheses?

Reply:
fixed